sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -533,15 +559,15 @@ object TypeCoercion {
       case a @ CreateArray(children) if !haveSameType(children) =>
         val types = children.map(_.dataType)
         findWiderCommonType(types) match {
-          case Some(finalDataType) => CreateArray(children.map(Cast(_, finalDataType)))
+          case Some(finalDataType) => CreateArray(children.map(castIfNotSameType(_, finalDataType)))
```

Comment:
is it needed? I think optimizer can remove unnecessary cast.

Reply:
Currently the optimizer doesn't remove the cast when the difference of the `finalDataType` is only the nullabilities of nested types.